File: mockserver/mockserver-core/src/main/java/org/mockserver/load/LoadScenario.java
package org.mockserver.load;

import org.mockserver.model.HttpTemplate;
import org.mockserver.model.ObjectWithJsonToString;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * An API-driven load scenario: an ordered list of templated request {@link LoadStep}s
 * driven at a target concurrency described by a {@link LoadProfile}, with optional
 * per-iteration data variation via the scenario's {@link HttpTemplate.TemplateType}.
 *
 * <p>Modelled on {@link org.mockserver.verify.VerificationSequence} (an ordered list of
 * request steps) but oriented at <em>producing</em> traffic rather than asserting over it.
 * It is a pure SLI producer: it records latency/error samples (into the metrics histograms
 * and the SLO sample store) but contains no verdict logic — the SLO verify feature consumes
 * those samples.
 *
 * <p>Off by default; {@code PUT /mockserver/loadScenario} returns 403 unless
 * {@code loadGenerationEnabled} is set.
 */
public class LoadScenario extends ObjectWithJsonToString {

    /**
     * How each iteration selects which steps to run from {@link #getSteps()}.
     */
    public enum StepSelection {
        /**
         * Default: every iteration runs ALL steps in declared order (a multi-step user journey).
         * Per-step {@link LoadStep#getWeight() weights} are ignored.
         */
        SEQUENTIAL,
        /**
         * Each iteration runs exactly ONE step, chosen at random with probability proportional to the
         * steps' {@link LoadStep#getWeight() weights} (mixed-workload modelling, e.g. 70% browse /
         * 20% search / 10% checkout). Because a WEIGHTED iteration runs a single step, cross-step
         * captures have no later step to consume them and so are meaningful only under SEQUENTIAL;
         * feeder data and pacing apply to both modes.
         */
        WEIGHTED
    }

    private String name;
    private List<LoadStep> steps = new ArrayList<>();
    private LoadProfile profile;
    private HttpTemplate.TemplateType templateType = HttpTemplate.TemplateType.VELOCITY;
    private Integer maxRequests;
    /**
     * Delay in milliseconds, applied after this scenario is <em>triggered</em> to start, before its
     * iterations begin. A scenario with {@code startDelayMillis > 0} sits in the {@code PENDING}
     * lifecycle state until the delay elapses (measured against the orchestrator's clock), then
     * transitions to {@code RUNNING}. {@code 0} (the default) means start immediately on trigger.
     * Lets several preloaded scenarios be triggered together yet begin at staggered offsets.
     */
    private long startDelayMillis;
    /**
     * Scenario-level custom annotation labels. Attached as OpenTelemetry attributes on every
     * load measurement (arbitrary keys — this is the flexible, k6-beating annotation surface) and
     * surfaced in the status DTO. For the Prometheus side, only the keys named in the
     * {@code mockserver.loadGenerationMetricLabels} allowlist are added as extra fixed labels
     * (Prometheus requires a fixed label-name set). Step labels (see {@link LoadStep#getLabels()})
     * are merged on top of these for a given step, with step keys winning on conflict.
     */
    private Map<String, String> labels;
    /**
     * In-run pass/fail thresholds (nullable/empty = none). Evaluated on every control tick from this
     * run's per-run latency histogram and counters; the run carries a {@code PASS} verdict iff ALL
     * thresholds hold, {@code FAIL} otherwise (and a null verdict until the first request completes,
     * or always-null when there are no thresholds). See {@link LoadThreshold}.
     */
    private List<LoadThreshold> thresholds;
    /**
     * When true, a {@code FAIL} verdict aborts the run early — the run transitions to the terminal
     * {@code STOPPED} state and is marked aborted-by-threshold. Default {@code false} (the run always
     * finishes its stages and carries a final verdict). See {@link #abortGraceMillis}.
     */
    private boolean abortOnFail;
    /**
     * Suppress {@code abortOnFail} for the first {@code N} milliseconds of the run, so noisy
     * startup samples cannot trigger a premature abort (analogous to k6's {@code delayAbortEval}).
     * Thresholds are still evaluated and a verdict still reported during the grace window — only the
     * abort action is deferred until {@code elapsedMillis >= abortGraceMillis}. Default {@code 0}.
     */
    private long abortGraceMillis;
    /**
     * Adaptive iteration pacing (think-time): a target per-VU iteration cycle time. Nullable (or
     * {@link LoadPacing.Mode#NONE}) means no pacing — the closed-model VU loop reschedules each next
     * iteration immediately. When set, a closed-model VU starts an iteration at most once per target
     * cycle (waiting out any remainder, starting immediately on overrun). Applies only to the
     * closed-model VU loop; open-model RATE iterations ignore it. See {@link LoadPacing}.
     */
    private LoadPacing pacing;
    /**
     * Parameterized test data ("data feeder"): an inline dataset from which the orchestrator selects
     * one row per iteration and exposes it to that iteration's templated request fields (path, body and
     * headers) as {@code $iteration.data.<column>} / {@code {{iteration.data.<column>}}}. Nullable
     * (default) means no feeder — {@code $iteration.data} resolves to an empty map and behaviour is
     * unchanged. With {@link LoadFeeder.Strategy#SEQUENTIAL SEQUENTIAL} selection the run completes once
     * the dataset is exhausted (each row used exactly once). See {@link LoadFeeder}.
     */
    private LoadFeeder feeder;
    /**
     * Per-iteration step selection mode (nullable = {@link StepSelection#SEQUENTIAL SEQUENTIAL}). Under
     * SEQUENTIAL (the default and the original behaviour) every iteration runs all steps in order; under
     * {@link StepSelection#WEIGHTED WEIGHTED} each iteration runs a single step chosen by
     * {@link LoadStep#getWeight() weight}. See {@link StepSelection}.
     */
    private StepSelection stepSelection;

    public static LoadScenario loadScenario() {
        return new LoadScenario();
    }

    public static LoadScenario loadScenario(String name) {
        return new LoadScenario().withName(name);
    }

    public String getName() {
        return name;
    }

    public LoadScenario withName(String name) {
        this.name = name;
        return this;
    }

    public List<LoadStep> getSteps() {
        return steps;
    }

    public LoadScenario withSteps(LoadStep... steps) {
        Collections.addAll(this.steps, steps);
        return this;
    }

    public LoadScenario withSteps(List<LoadStep> steps) {
        this.steps = steps;
        return this;
    }

    public LoadProfile getProfile() {
        return profile;
    }

    public LoadScenario withProfile(LoadProfile profile) {
        this.profile = profile;
        return this;
    }

    public HttpTemplate.TemplateType getTemplateType() {
        return templateType;
    }

    public LoadScenario withTemplateType(HttpTemplate.TemplateType templateType) {
        this.templateType = templateType;
        return this;
    }

    public Integer getMaxRequests() {
        return maxRequests;
    }

    public LoadScenario withMaxRequests(Integer maxRequests) {
        this.maxRequests = maxRequests;
        return this;
    }

    /**
     * The start delay in milliseconds applied after a start trigger before this scenario's
     * iterations begin (default {@code 0}). See {@link #startDelayMillis}.
     */
    public long getStartDelayMillis() {
        return startDelayMillis;
    }

    public LoadScenario withStartDelayMillis(long startDelayMillis) {
        this.startDelayMillis = startDelayMillis;
        return this;
    }

    /**
     * Scenario-level custom labels (may be null/empty). See {@link #labels}.
     */
    public Map<String, String> getLabels() {
        return labels;
    }

    public LoadScenario withLabels(Map<String, String> labels) {
        this.labels = labels != null ? new LinkedHashMap<>(labels) : null;
        return this;
    }

    public LoadScenario withLabel(String name, String value) {
        if (this.labels == null) {
            this.labels = new LinkedHashMap<>();
        }
        this.labels.put(name, value);
        return this;
    }

    /**
     * In-run pass/fail thresholds (may be null/empty). See {@link #thresholds}.
     */
    public List<LoadThreshold> getThresholds() {
        return thresholds;
    }

    public LoadScenario withThresholds(LoadThreshold... thresholds) {
        if (this.thresholds == null) {
            this.thresholds = new ArrayList<>();
        }
        Collections.addAll(this.thresholds, thresholds);
        return this;
    }

    public LoadScenario withThresholds(List<LoadThreshold> thresholds) {
        this.thresholds = thresholds;
        return this;
    }

    /**
     * Whether a {@code FAIL} verdict aborts the run early (default {@code false}). See {@link #abortOnFail}.
     */
    public boolean isAbortOnFail() {
        return abortOnFail;
    }

    public LoadScenario withAbortOnFail(boolean abortOnFail) {
        this.abortOnFail = abortOnFail;
        return this;
    }

    /**
     * The abort grace window in milliseconds (default {@code 0}). See {@link #abortGraceMillis}.
     */
    public long getAbortGraceMillis() {
        return abortGraceMillis;
    }

    public LoadScenario withAbortGraceMillis(long abortGraceMillis) {
        this.abortGraceMillis = abortGraceMillis;
        return this;
    }

    /**
     * Adaptive iteration pacing (may be null = no pacing). See {@link #pacing}.
     */
    public LoadPacing getPacing() {
        return pacing;
    }

    public LoadScenario withPacing(LoadPacing pacing) {
        this.pacing = pacing;
        return this;
    }

    /**
     * The data feeder (may be null = no feeder). See {@link #feeder}.
     */
    public LoadFeeder getFeeder() {
        return feeder;
    }

    public LoadScenario withFeeder(LoadFeeder feeder) {
        this.feeder = feeder;
        return this;
    }

    /**
     * The per-iteration step selection mode (may be null = {@link StepSelection#SEQUENTIAL}). See
     * {@link #stepSelection}.
     */
    public StepSelection getStepSelection() {
        return stepSelection;
    }

    public LoadScenario withStepSelection(StepSelection stepSelection) {
        this.stepSelection = stepSelection;
        return this;
    }
}

File: mockserver-client-php/src/LoadScenario.php
<?php

declare(strict_types=1);

namespace MockServer;

/**
 * A small fluent value object describing an API load scenario for
 * {@see MockServerClient::loadScenario()}.
 *
 * A scenario is an ordered list of templated request steps fired at a target
 * load described by a staged {@see LoadProfile}, with per-iteration data
 * variation. It is a pure SLI producer: each completed request records a
 * latency/error sample that {@see MockServerClient::verifySlo()} can read.
 *
 * Each scenario carries a unique {@code name} that the server's registry keys it
 * by. {@see MockServerClient::loadScenario()} only registers (loads) the
 * scenario; {@see MockServerClient::startLoadScenarios()} drives the load.
 *
 * This object is optional — {@see MockServerClient::loadScenario()} also accepts
 * a plain associative array in the same camelCase shape — but it gives a typed,
 * discoverable builder that emits the correct JSON.
 *
 * @example
 *   $scenario = LoadScenario::scenario('checkout-load')
 *       ->maxRequests(5000)
 *       ->profile(LoadProfile::of(
 *           LoadStage::vuRamp(1, 10, 10000),
 *           LoadStage::vuHold(10, 30000),
 *       ))
 *       ->addStep(
 *           HttpRequest::request()->method('GET')->path('/api/item/$iteration.index'),
 *           Delay::milliseconds(20),
 *       );
 *   $client->loadScenario($scenario);
 */
class LoadScenario implements \JsonSerializable
{
    private string $name;
    private ?string $templateType = null;
    private ?int $maxRequests = null;
    private ?int $startDelayMillis = null;
    private ?LoadProfile $profile = null;
    /** @var array<int, array<string, mixed>> */
    private array $steps = [];
    /** @var array<string, string> */
    private array $labels = [];
    /** @var array<int, LoadThreshold> */
    private array $thresholds = [];
    private ?bool $abortOnFail = null;
    private ?int $abortGraceMillis = null;
    private ?LoadPacing $pacing = null;
    private ?LoadFeeder $feeder = null;
    private ?string $stepSelection = null;

    public function __construct(string $name)
    {
        $this->name = $name;
    }

    /**
     * Begin building a scenario with the given human-readable name.
     */
    public static function scenario(string $name): self
    {
        return new self($name);
    }

    /**
     * The unique name this scenario is registered under.
     */
    public function getName(): string
    {
        return $this->name;
    }

    /**
     * Set the template engine used to render per-iteration paths and bodies.
     *
     * Only VELOCITY (default) and MUSTACHE are supported for load steps.
     */
    public function templateType(string $templateType): self
    {
        $this->templateType = strtoupper($templateType);

        return $this;
    }

    /**
     * Set a hard cap on the total number of requests dispatched.
     */
    public function maxRequests(int $maxRequests): self
    {
        $this->maxRequests = $maxRequests;

        return $this;
    }

    /**
     * Set an optional delay (in milliseconds) applied after the scenario is
     * started before it begins driving load. Honoured by
     * {@see MockServerClient::startLoadScenarios()}.
     */
    public function startDelayMillis(int $startDelayMillis): self
    {
        $this->startDelayMillis = $startDelayMillis;

        return $this;
    }

    /**
     * Set the staged load profile describing the load over time.
     */
    public function profile(LoadProfile $profile): self
    {
        $this->profile = $profile;

        return $this;
    }

    /**
     * Add a templated request step.
     *
     * @param HttpRequest $request The (templated) request to fire each iteration
     * @param Delay|null $thinkTime Optional inter-step pause
     * @param string|null $name Optional human label / metric label for the step
     * @param array<string, string> $labels Optional step-level annotation labels
     * @param array<int, LoadCapture> $captures Optional cross-step capture rules
     *        applied to this step's response (visible to subsequent steps in the
     *        same iteration).
     * @param float|null $weight Relative selection weight, used only when the
     *        scenario's stepSelection is WEIGHTED.
     */
    public function addStep(
        HttpRequest $request,
        ?Delay $thinkTime = null,
        ?string $name = null,
        array $labels = [],
        array $captures = [],
        ?float $weight = null,
    ): self {
        $step = ['request' => $request->toArray()];
        if ($thinkTime !== null) {
            $step['thinkTime'] = $thinkTime->toArray();
        }
        if ($name !== null) {
            $step['name'] = $name;
        }
        if ($labels !== []) {
            $step['labels'] = $labels;
        }
        if ($captures !== []) {
            $step['captures'] = array_map(
                static fn (LoadCapture $capture): array => $capture->toArray(),
                array_values($captures),
            );
        }
        if ($weight !== null) {
            $step['weight'] = $weight;
        }
        $this->steps[] = $step;

        return $this;
    }

    /**
     * Set scenario-level annotation labels (exported as OpenTelemetry attributes
     * and, for allowlisted keys, Prometheus labels).
     *
     * @param array<string, string> $labels
     */
    public function labels(array $labels): self
    {
        $this->labels = $labels;

        return $this;
    }

    /**
     * Set the in-run pass/fail thresholds. The run carries a PASS verdict iff
     * all hold, FAIL otherwise. Empty means no verdict is computed.
     *
     * @param array<int, LoadThreshold> $thresholds
     */
    public function thresholds(LoadThreshold ...$thresholds): self
    {
        $this->thresholds = array_values($thresholds);

        return $this;
    }

    /**
     * Append a single in-run pass/fail threshold.
     */
    public function addThreshold(LoadThreshold $threshold): self
    {
        $this->thresholds[] = $threshold;

        return $this;
    }

    /**
     * When true, a FAIL verdict aborts the run early (terminal STOPPED state,
     * abortedByThreshold set); default false (the run always finishes its
     * stages and carries a final verdict).
     */
    public function abortOnFail(bool $abortOnFail = true): self
    {
        $this->abortOnFail = $abortOnFail;

        return $this;
    }

    /**
     * Suppress abortOnFail for the first N milliseconds of the run so noisy
     * startup samples cannot trigger a premature abort.
     */
    public function abortGraceMillis(int $abortGraceMillis): self
    {
        $this->abortGraceMillis = $abortGraceMillis;

        return $this;
    }

    /**
     * Set adaptive iteration pacing (think-time) for the closed-model VU loop.
     */
    public function pacing(LoadPacing $pacing): self
    {
        $this->pacing = $pacing;

        return $this;
    }

    /**
     * Set the parameterized test data (data feeder) for the scenario.
     */
    public function feeder(LoadFeeder $feeder): self
    {
        $this->feeder = $feeder;

        return $this;
    }

    /**
     * Set how each iteration selects which steps to run: SEQUENTIAL (default,
     * all steps in declared order) or WEIGHTED (exactly one step per iteration
     * chosen at random proportional to each step's weight).
     */
    public function stepSelection(string $stepSelection): self
    {
        $this->stepSelection = strtoupper($stepSelection);

        return $this;
    }

    /**
     * @return array<string, mixed>
     */
    public function jsonSerialize(): array
    {
        return $this->toArray();
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $result = ['name' => $this->name];
        if ($this->templateType !== null) {
            $result['templateType'] = $this->templateType;
        }
        if ($this->maxRequests !== null) {
            $result['maxRequests'] = $this->maxRequests;
        }
        if ($this->startDelayMillis !== null) {
            $result['startDelayMillis'] = $this->startDelayMillis;
        }
        if ($this->labels !== []) {
            $result['labels'] = $this->labels;
        }
        if ($this->thresholds !== []) {
            $result['thresholds'] = array_map(
                static fn (LoadThreshold $threshold): array => $threshold->toArray(),
                $this->thresholds,
            );
        }
        if ($this->abortOnFail !== null) {
            $result['abortOnFail'] = $this->abortOnFail;
        }
        if ($this->abortGraceMillis !== null) {
            $result['abortGraceMillis'] = $this->abortGraceMillis;
        }
        if ($this->pacing !== null) {
            $result['pacing'] = $this->pacing->toArray();
        }
        if ($this->feeder !== null) {
            $result['feeder'] = $this->feeder->toArray();
        }
        if ($this->profile !== null) {
            $result['profile'] = $this->profile->toArray();
        }
        if ($this->stepSelection !== null) {
            $result['stepSelection'] = $this->stepSelection;
        }
        $result['steps'] = $this->steps;

        return $result;
    }
}
